### PR TITLE
Fix merge issues with ZoL/master

### PR DESCRIPTION
**Pull request by pcd1193182**
This commit merges in the latest commits from ZoL master, resolving some send/recv related merge conflicts in the process.

### CODE REVIEW

**Comment by ahrens** [comment]
Can you post the output of `git show <merge commit>`, which should show us what the merge conflicts were and how you resolved them?

**Comment by ahrens** [comment]
Realized I could do this myself:
```
$ git show origin/156/head
commit b33a06ca6cb3430214dac0e7d62e3e23063de611 (origin/156/head)
Merge: e87eb86210 c14ca1456e
Author: Paul Dagnelie <pcd@delphix.com>
Date:   Tue Apr 28 16:28:39 2020 -0700

    Fix merge issues with ZoL master

diff --cc lib/libzfs_core/libzfs_core.c
index 76fc4ddad7,4e83b624b2..5a25bcbc8f
--- a/lib/libzfs_core/libzfs_core.c
+++ b/lib/libzfs_core/libzfs_core.c
@@@ -982,22 -980,7 +982,22 @@@ lzc_receive_resumable(const char *snapn
      boolean_t force, boolean_t raw, int fd)
  {
        return (recv_impl(snapname, props, NULL, NULL, 0, origin, force,
-           B_TRUE, raw, B_FALSE, fd, NULL, NULL, NULL, NULL));
 -          B_TRUE, raw, fd, NULL, NULL, NULL, NULL));
++          B_TRUE, raw, B_FALSE, , NULL, NULL, NULL, NULL));
 +}
 +
 +/*
 + * Like lzc_receive_resumable, but DRR_OBJECT_RANGE records will be treated as
 + * REDACT records, and REDACT records will cause an assertion. This is
 + * intended to resolve an issue where streams from DxOS systems have the two
 + * reversed because that OS had redaction land before encryption, unlike all
 + * other platforms.
 + */
 +int
 +lzc_receive_resumable_illumos(const char *snapname, nvlist_t *props,
 +    const char *origin, boolean_t force, boolean_t raw, int fd)
 +{
 +      return (recv_impl(snapname, props, NULL, NULL, 0, origin, force,
 +          B_TRUE, raw, B_FALSE, fd, NULL, NULL, NULL, NULL));
  }

  /*
diff --cc module/zfs/zfs_ioctl.c
index 7ec9367033,2104bef714..c7f2b81e6b
--- a/module/zfs/zfs_ioctl.c
+++ b/module/zfs/zfs_ioctl.c
@@@ -5237,8 -5236,7 +5237,7 @@@ zfs_ioc_recv_new(const char *fsname, nv

        force = nvlist_exists(innvl, "force");
        resumable = nvlist_exists(innvl, "resumable");
 -
 +      illumos = nvlist_exists(innvl, "illumos");
-
        /* we still use "props" here for backwards compatibility */
        error = nvlist_lookup_nvlist(innvl, "props", &recvprops);
        if (error && error != ENOENT)
```

File: lib/libzfs_core/libzfs_core.c [hunk 1]
```diff
@@ -982,7 +982,7 @@ lzc_receive_resumable(const char *snapname, nvlist_t *props, const char *origin,
     boolean_t force, boolean_t raw, int fd)
 {
 	return (recv_impl(snapname, props, NULL, NULL, 0, origin, force,
-	    B_TRUE, raw, B_FALSE, fd, NULL, NULL, NULL, NULL));
+	    B_TRUE, raw, B_FALSE, , NULL, NULL, NULL, NULL));
```
Comment (ahrens) [inline lib/libzfs_core/libzfs_core.c:985]:
```suggestion
	    B_TRUE, raw, B_FALSE, NULL, NULL, NULL, NULL));
```

**Comment by ahrens** [comment]
Could you run `git ab-pre-push` and `git zfs-precommit`?

**Comment by ahrens** [comment]
I think you'll also need to remove the `fd` argument in the call to `recv_impl()` in `lzc_receive_resumable_illumos()` (github won't let me comment on a line that you didn't change)

**Comment by ahrens** [comment]
I thought that the "remove dedup code" was merged by https://github.com/delphix/zfs/pull/155 ?

**Comment by pcd1193182** [comment]
I'm not sure why we have to re-merge it, but this is what git appears to believe. I've fixed the compilation errors, waiting for a passing ab-pre-push and zfs-precommit run to post the new version.

**Comment by ahrens** [comment]
It looks like it was cherry-picked, rather than merged (perhaps because it was integrated with the "rebase and merge" button on the github PR).  Merges with upstream should always be created with `git merge`, and pushed with `git push [...]` (not the github UI).
```
* commit e87eb86210f5c5d991634262f6b683b3a3a4d830 (origin/master)
| Author: Don Brady <don.brady@delphix.com>
| Date:   Tue Apr 28 15:19:26 2020 -0600
|
|     Remove deduplicated send/receive code (#155)
```
 I'd recommend that we have one merge commit that does the commits other than  "Remove deduplicated send/receive code", and then have another merge commit which merges just that commit, with no diffs.

**Comment by pcd1193182** [comment]
It looks like the only merge conflicts come from that single commit, so I'm just going to change this review to be merging in that commit properly, and then let the automation merge the other ones.

**Comment by pcd1193182** [comment]
http://platform.jenkins.delphix.com/job/devops-gate/job/master/job/zfs-precommit/5188/console zloop failure appears to be a test bug.

**Pull request merged**